File: source/jugglinglab/jml/PathLink.java
// PathLink.java
//
// Copyright 2019 by Jack Boyce (jboyce@gmail.com)

package jugglinglab.jml;

import jugglinglab.path.*;
import jugglinglab.util.*;


public class PathLink {
    protected int           pathnum;
    protected JMLEvent      startevent, endevent;
    protected int           catchtype;
    protected String        throwtype, mod;
    protected Path          proppath;
    protected boolean       inhand;
    protected int           juggler, hand;

    protected boolean       ismaster;
    protected PathLink[]    duplicates;     // if master
    protected PathLink      master;         // if duplicate


    public PathLink(int pathnum, JMLEvent startevent, JMLEvent endevent) {
        this.pathnum = pathnum;
        this.startevent = startevent;
        this.endevent = endevent;
        this.proppath = null;
        this.inhand = false;
    }

    public void setThrow(String type, String mod) throws JuggleExceptionUser, JuggleExceptionInternal {
        proppath = Path.newPath(type);
        proppath.initPath(mod);
        proppath.setStart(startevent.getGlobalCoordinate(), startevent.getT());
        proppath.setEnd(endevent.getGlobalCoordinate(), endevent.getT());
        proppath.calcPath();
        this.throwtype = type;
        this.mod = mod;
        this.inhand = false;
    }

    public void setInHand(int juggler, int hand) {
        this.inhand = true;
        this.juggler = juggler;
        this.hand = hand;
    }

    public Path getPath()           { return proppath; }

    public int getCatch()           { return catchtype; }
    public void setCatch(int catchtype) { this.catchtype = catchtype; }
    public int getPathNum()         { return pathnum; }
    public JMLEvent getStartEvent() { return startevent; }
    public JMLEvent getEndEvent()   { return endevent; }

    public boolean isInHand()       { return inhand; }
    public int getHoldingJuggler()  { return juggler; }
    public int getHoldingHand()     { return hand; }
    public boolean isMaster()       { return ismaster; }

    @Override
    public String toString() {
        String result = null;

        if (inhand) {
            result = "In hand, ";
        } else {
            result = "Not in hand (type=\""+throwtype+"\", mod=\""+
            mod+"\"), ";
        }

        Coordinate start = startevent.getGlobalCoordinate();
        result += "from (x="+start.x+",y="+start.y+",z="+start.z+",t="+
            startevent.getT()+") ";
        Coordinate end = endevent.getGlobalCoordinate();
        result += "to (x="+end.x+",y="+end.y+",z="+end.z+",t="+endevent.getT()+")";
        return result;
    }
}

File: composeApp/src/commonMain/kotlin/jugglinglab/jml/PathLink.kt
//
// PathLink.kt
//
// Copyright 2002-2025 Jack Boyce and the Juggling Lab contributors
//

package jugglinglab.jml

import jugglinglab.path.Path
import jugglinglab.util.Coordinate
import jugglinglab.util.JuggleExceptionInternal
import jugglinglab.util.JuggleExceptionUser

class PathLink(
    val pathNum: Int,
    val startGlobalCoordinate: Coordinate,
    val startEvent: JMLEvent,
    val endGlobalCoordinate: Coordinate,
    val endEvent: JMLEvent
) {
    var path: Path? = null
        private set
    var isInHand: Boolean = false
        private set

    // for paths corresponding to a throw (isInHand false):
    var throwType: String? = null
        private set
    var throwMod: String? = null
        private set

    // for paths corresponding to a carry (isInHand true):
    var holdingJuggler: Int = 0
        private set
    var holdingHand: Int = 0
        private set

    @Throws(JuggleExceptionUser::class, JuggleExceptionInternal::class)
    fun setThrow(
        pathType: String,
        pathMod: String?
    ) {
        val newPath = Path.newPath(pathType)
        newPath.initPath(pathMod)
        newPath.setStart(startGlobalCoordinate, startEvent.t)
        newPath.setEnd(endGlobalCoordinate, endEvent.t)
        newPath.calcPath()
        path = newPath
        throwType = pathType
        throwMod = pathMod
        isInHand = false
    }

    fun setInHand(juggler: Int, hand: Int) {
        isInHand = true
        holdingJuggler = juggler
        holdingHand = hand
    }

    override fun toString(): String {
        val start = startGlobalCoordinate
        val end = endGlobalCoordinate
        val sb = StringBuilder()
        sb.append(
            if (isInHand) "In hand, " else
                "Not in hand (type=\"$throwType\", mod=\"$throwMod\"), "
        )
        sb.append("from (x=${start.x},y=${start.y},z=${start.z},t=${startEvent.t}) ")
        sb.append("to (x=${end.x},y=${end.y},z=${end.z},t=${endEvent.t})")
        return sb.toString()
    }
}
